### ISSUE TITLE
[Clarifications] Correct way to re-run the web application without losing previously collected data

### ISSUE THREAD

**Issue by federicoruggeri** [facebookresearch/Mephisto#429]
Dear All,
Apologies for the very stupid question, but I'm afraid I'm not getting it right.

I've noticed that re-running the same application (from the main `run.py` script) always gets new tasks (assignment ids) without the possibility to maintain previous ones (I'm using a custom ParlAI Teacher under the hood).

For example, If I run the dialogue-based application with `num_conversations`  equal to 1, I initially get `assignment_id` equal to 1 and 2. Subsequently, if I stop and re-run the application, I'll get `assignment_id` equal to 3 and 4.

Under Heroku architect, things get a little bit more complicated because I can't re-run `run.py` unless I delete the application on my Heroku account webpage. After this operation, I lose all acquired data from previous tasks, while maintaining onboarded workers in the database.

For example, If we call the first run of the application **Run 1** with worker **worker_1**. Then, after deleting and re-running the application (**Run 2**), I can see **worker_1** within the database but its content is now empty.

I'm not sure if data gets overwritten or if the data from previous runs is still there and the read operation with the same task name points now to a different location within the local database.

------

To sum up, what is the best practice to run the application multiple times with the capability of controlling which tasks are being considered? For example, I would like to run the application, do some tests, collect that data, shut down the application, modify the code and then re-run with the same assignment ids but without losing previously gathered test data.

Additionally, what's the correct way to re-run Heroku based applications? Should I re-start the app from the Heroku CLI?

Thanks in advance for your time and courtesy!

**Comment by JackUrb** [comment]
Hi @federicoruggeri - a few points here:

> I've noticed that re-running the same application (from the main run.py script) always gets new tasks (assignment ids) without the possibility to maintain previous ones (I'm using a custom ParlAI Teacher under the hood).

Correct, this is part of how Mephisto is implemented at the moment. On our roadmap is a feature to be able to "relaunch a copy" or "relaunch incomplete" from a previous task, but we haven't gotten this together yet. At the moment we ask that you do local bookkeeping of some kind, such that you can track which tasks to launch with on future runs.

> I can't re-run run.py unless I delete the application on my Heroku account webpage.

The `HerokuArchitect` should go through a normal shutdown procedure, provided you aren't killing the task abruptly. Shutdown is triggered either automatically when all launched tasks are completed, or by a single Ctrl-C interruption. After cleaning up all `Unit`s that aren't in-flight, Mephisto may pause to ensure that that any `Unit` currently being worked on has the required time to complete. A second Ctrl-C interruption can be used here to skip the wait period, after which the server should close.

> After this operation, I lose all acquired data from previous tasks, while maintaining onboarded workers in the database.

Data is stored by `Agent`, not by `Worker`. `Agent`s are a pairing between a `Worker` and `Unit` that allows us to uniquely index work, and store different types of work. You can query your results with the `MephistoDataBrowser`, and an example of how to do this is in this [`examine_results.py` script](https://github.com/facebookresearch/Mephisto/blob/master/examples/simple_static_task/examine_results.py).

Let me know if there's more here that needs explaining!

**Comment by federicoruggeri** [comment]
That's perfect, thank you very much for your quick response!
